sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala

```diff
@@ -217,4 +218,12 @@ class ConstraintPropagationSuite extends SparkFunSuite {
         IsNotNull(resolveColumn(tr, "a")),
         IsNotNull(resolveColumn(tr, "b")))))
   }
+
+  test("infer constraints on cast") {
+    val tr = LocalRelation('a.int, 'b.long)
+    verifyConstraints(tr.where('a.attr === 'b.attr).analyze.constraints,
```

Comment:
Since your implementation also handles nested Casts, can we add an explicit test for that? Perhaps the filter condition can be `'a.attr === 'b.attr && IsNotNull(Cast(Cast(resolveColumn(tr, "a"), LongType), LongType)))`